### ISSUE TITLE
Default value for Callable fails the type check, accepted when passed explicitly

### ISSUE THREAD

**Issue by tlandschoff-scale** [python/mypy#15012]
**Bug Report**

tl;dr: mypy rejects a default argument for a callable parameter that is accepted when passed directly.

Basically:

```python
def works(reducer: Callable[[Input], T]) -> T: ...
reveal_type(works(lambda x: sum(x)))  # Output: Revealed type is "builtins.int"

def doesnt(reducer: Callable[[Input], T] = lambda x: sum(x)) -> T: ...
reveal_type(doesnt())  # Output: note: Revealed type is "<nothing>"
```

mypy also complains about the default argument having an incompatible type (unbound `T` appears to be incompatible with `int` :shrug:):

```
Incompatible default for argument "reducer" (default has type "Callable[[List[int]], int]", argument has type "Callable[[List[int]], T]")  [assignment]
```

**To Reproduce**

I pasted a minimal, runnable example to the [playground](https://mypy-play.net/?mypy=master&python=3.11&gist=44f20070e9b7e7e0a4ddd16713822efe). In case it is down, here is the [gist](https://gist.github.com/mypy-play/44f20070e9b7e7e0a4ddd16713822efe).

This was stripped down to the bare minimum to show the issue. A more reasonable example (still stripped down) would be [this (playground)](https://mypy-play.net/?mypy=master&python=3.11&gist=31fa0cc5a2ee35a1d96e867eedf2ea5e), for [this gist](https://gist.github.com/mypy-play/31fa0cc5a2ee35a1d96e867eedf2ea5e).

**Expected Behavior**

I would have expected that mypy comes to the same conclusion as myself and `pyright`:

```console
$ pyright trivial.py 
pyright 1.1.257
/home/torsten.landschoff/trivial.py
  /home/torsten.landschoff/trivial.py:9:13 - information: Type of "works(lambda x: sum(x))" is "int"
  /home/torsten.landschoff/trivial.py:14:13 - information: Type of "doesnt()" is "int"
0 errors, 0 warnings, 2 informations 
Completed in 0.788sec
```

For the less trivial example which is a bit closer to my actual source code, pyright reports:

```console
$ pyright less_trivial.py 
pyright 1.1.257
/home/torsten.landschoff/less_trivial.py
  /home/torsten.landschoff/less_trivial.py:27:13 - information: Type of "WorkingCache().perform_query(Query(Customer), lambda x: x)" is "Iterable[Customer]"
  /home/torsten.landschoff/less_trivial.py:30:13 - information: Type of "WorkingCache().perform_query(Query(Customer), lambda results: Counter(c.home_town for c in results))" is "Counter[str]"
  /home/torsten.landschoff/less_trivial.py:44:13 - information: Type of "BrokenCache().perform_query(Query(Customer))" is "Iterable[Customer]"
0 errors, 0 warnings, 3 informations 
Completed in 0.893sec
```

**Actual Behavior**

```console
$ mypy --version
mypy 1.3.0+dev.e21ddbf3ebe80bc219708a5f69a1b7d0b4e20400 (compiled: no)
$ mypy trivial.py 
trivial.py:9: note: Revealed type is "builtins.int"
trivial.py:11: error: Incompatible default for argument "reducer" (default has type "Callable[[List[int]], int]", argument has type "Callable[[List[int]], T]")  [assignment]
trivial.py:11: error: Incompatible return value type (got "int", expected "T")  [return-value]
trivial.py:14: note: Revealed type is "<nothing>"
Found 2 errors in 1 file (checked 1 source file)
```

For the more complete example that actually showcases the idea behind the code, mypy reports these errors:

```console
$ mypy less_trivial.py 
less_trivial.py:27: note: Revealed type is "typing.Iterable[less_trivial.Customer]"
less_trivial.py:30: note: Revealed type is "collections.Counter[builtins.str]"
less_trivial.py:40: error: Incompatible default for argument "reducer" (default has type "Callable[[Iterable[T]], Iterable[T]]", argument has type "Callable[[Iterable[T]], U]")  [assignment]
less_trivial.py:40: error: Incompatible return value type (got "Iterable[T]", expected "U")  [return-value]
less_trivial.py:44: note: Revealed type is "<nothing>"
Found 2 errors in 1 file (checked 1 source file)
```

**Your Environment**

Not sure this is that relevant, given that the mypy-playground link includes that information. But here it is:

- Mypy version used: `mypy 1.3.0+dev.e21ddbf3ebe80bc219708a5f69a1b7d0b4e20400 (compiled: no)`
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): Not using any configuration file
- Python version used: `Python 3.12.0a7+`

**Comment by tlandschoff-scale** [comment]
FYI: I found a [workaround](https://mypy-play.net/?mypy=master&python=3.11&gist=67fd10775e7d6cc1399242681d46e915) ([gist](https://gist.github.com/mypy-play/67fd10775e7d6cc1399242681d46e915)) using `overload` to make this more explicit.

Basically:

```python
@overload
def workaround() -> int:
    ...

@overload
def workaround(reducer: Callable[[Input], T]) -> T:
    ...

def workaround(reducer: Optional[Callable[[Input], T]] = None) -> T:
    ...
```

I am not particularly happy with this, though.

**Comment by hauntsaninja** [comment]
Thanks, I think this is a essentially a duplicate of #3737

**Comment by tlandschoff-scale** [comment]
@hauntsaninja Thanks for looking at this. I agree, this is covered by #3737.
